src/Property.php

```diff
@@ -105,6 +105,16 @@ protected function isValidType($value): bool
 
     protected function assertTypeEquals(string $type, $value): bool
     {
+        if (is_array($value)) {
```

Comment:
With your approach, you're inferring the type "array of" based on the value. It should be the other way around: the type should be defined, and the value should follow that type. If not, a type error is thrown.

In practice you'll want to validate whether `$type` indicates an array. Say for example that `$type` is `string[]`:

```php
if (strpos('[]', $type) !== false)
{
    if (! is_array($value)) {
        return false;
    }

    $arrayType = str_replace('[]', '', $type);

    foreach ($value as $valueItem) {
        $this->assertTypeEquals($arrayType, $valueItem);
    }
}
```

(Be sure to double check whether I got the parameter ordening right in the `str` functions 😝)

Perhaps it would be good to split this functionality into a separate function, to keep `assertTypeEquals` cleaner. You can call that function in `assertTypeEquals` of course.

Reply:
That is a lot better indeed! I wasn't sure I could alter the `preg_match` usage in `resolveTypeDefinition`. After adding the `[]` it worked as expected.